virttest/storage.py

```diff
@@ -336,8 +336,12 @@ def image_secret_define_by_params(cls, image, params):
     def save_to_file(self):
         """Save secret data to file."""
         _make_secret_dir()
-        with open(self.filename, "w") as fd:
-            fd.write(self.data)
+        try:
+            with open(self.filename, "w") as fd:
+                fd.write(self.data)
+        except TypeError:
+            with open(self.filename, "wb") as fd:
```

Comment:
Would you please provide the error message you met? it should be either opened with or without the `b` flag, not just try one then fall back to another when the attempt failed.

Reply:
When create luks image with non_utf8_secret, hit this issue.
\# echo -n -e '\x3a\x3c\x3b\xff' > non_utf8_secret
# qemu-img create -f luks --object secret,id=sec0,file=non_utf8_secret -o key-secret=sec0 test.luks 4M

For non utf8 string, we have to store it as binary file. And then create object secrect by "--object secret,id={s.aid},file={s.filename}". So retry it with \"wb\".

Reply:
sorry, I don't quite understand your issue, what content do you want to store, bytes or texts?

Reply:
The content was utf8 string, and stored as texts by default.  Just for non utf8 string testing, hit TypeError when saved it as texts. So store it as a binary file.

TypeError: write() argument must be str, not bytes

Reply:
1. currently, the `data` attribute of the `ImageSecret` class only supports `str` values (although it is not documented), that means you have to restructure it if you want to support `bytes` values
2. `open` supports specifying encoding type on python3 ([ref](https://docs.python.org/3/library/functions.html#open)), which means you can get non-utf8 files without using binary mode, e.g.
```
$ cat test.py
with open('/tmp/secret', 'w', encoding='euc_jp') as f:
    f.write('\x3a\x3c\x3b\xff')
$ python3 test.py
$ qemu-img create -f luks --object secret,id=sec0,file=/tmp/secret -o key-secret=sec0 test.luks 4M
Formatting 'test.luks', fmt=luks size=4194304 key-secret=sec0
qemu-img: test.luks: Data from secret sec0 is not valid UTF-8
```
So, the point is, no matter what you want is 1 or 2, `ImageSecret` (or accurately `ImageSecret.data`) have to be restructured, instead of simply using the current approach.

Reply:
Thanks for your detailed explanation.
Since non-utf8 file is negative test, I think set a default parameter \"encoding\" to save_to_file is OK.
```
def save_to_file(self, encoding="utf-8"):
        """Save secret data to file."""
        _make_secret_dir()
        with open(self.filename, "w", encoding=encoding) as fd:
            fd.write(self.data)
```

And add a parameter \"image_secret_encoding_type\", in order to specify encoding type in test case.  
```
if self.encryption_config.key_secret:
     encoding = params.get(\"image_secret_encoding_type\", \"utf-8\")
     self.encryption_config.key_secret.save_to_file(encoding=encoding)
```

Do you agree with the above update? thanks